src/System.Security.Cryptography.Pkcs/src/System/Security/Cryptography/Pkcs/SignedCms.cs

```diff
@@ -152,6 +170,59 @@ internal void Decode(ReadOnlyMemory<byte> encodedMessage)
             _hasData = true;
         }
 
+        internal static ReadOnlyMemory<byte> GetContent(
+            ReadOnlyMemory<byte> wrappedContent,
+            string contentType)
+        {
+            // Read the input.
+            //
+            // PKCS7's id-data is written in both PKCS#7 and CMS as an OCTET STRING wrapping
+            // the arbitrary bytes, so the OCTET STRING must always be present.
+            //
+            // For other types, CMS says to always write an OCTET STRING, and to put the properly
+            // encoded data within it.
+            // PKCS#7 originally ommitted the OCTET STRING wrapper for this model, so this is the
+            // dynamic adapter.
+            //
+            // See https://tools.ietf.org/html/rfc5652#section-5.2.1
+            byte[] rented = null;
+            int bytesWritten = 0;
+            try
+            {
+                AsnReader reader = new AsnReader(wrappedContent, AsnEncodingRules.BER);
+
+                if (reader.TryGetPrimitiveOctetStringBytes(out ReadOnlyMemory<byte> inner))
+                {
+                    return inner;
+                }
+
+                rented = ArrayPool<byte>.Shared.Rent(wrappedContent.Length);
+
+                if (!reader.TryCopyOctetStringBytes(rented, out bytesWritten))
+                {
+                    Debug.Fail($"TryCopyOctetStringBytes failed with an array larger than the encoded value");
+                    throw new CryptographicException();
+                }
+
+                return rented.AsSpan(0, bytesWritten).ToArray();
+            }
+            catch (Exception) when (contentType != Oids.Pkcs7Data)
```

Comment:
Should this be CryptographicException?

Reply:
There shouldn't be any other errors there, but the guidance from the RFC is

> ... If the
   implementation is unable to ASN.1 decode the SignedData type using
   the CMS SignedData encapContentInfo eContent OCTET STRING syntax,
   then the implementation MAY attempt to decode the SignedData type
   using the PKCS #7 SignedData contentInfo content ANY syntax and
   compute the message digest accordingly.

So I went with "if anything goes wrong, fall back".

Reply:
fair enough